enterprise/server/remote_execution/containers/firecracker/firecracker.go

```diff
@@ -2550,6 +2580,70 @@ func (c *FirecrackerContainer) pause(ctx context.Context) error {
 	return nil
 }
 
+// reclaimMemoryWithBalloon attempts to decrease memory snapshot size by expanding
+// the memory balloon to 90% of available memory.
+func (c *FirecrackerContainer) reclaimMemoryWithBalloon(ctx context.Context) error {
+	conn, err := c.vmExecConn(ctx)
+	if err != nil {
+		return status.InternalErrorf("failed to dial VM exec port: %s", err)
+	}
+	client := vmxpb.NewExecClient(conn)
+	cmd := &repb.Command{
+		Arguments: []string{"sh", "-c", "free | awk '/^Mem:/ {print $7}'"},
+	}
+	res := vmexec_client.Execute(ctx, client, cmd, "/workspace/", c.user, nil /* statsListener*/, &interfaces.Stdio{})
+	if res.Error != nil {
+		return res.Error
+	}
+	availableMemKB, err := strconv.Atoi(strings.TrimSpace(string(res.Stdout)))
+	if err != nil {
+		return status.WrapErrorf(err, "parse available mem: %s", string(res.Stdout))
+	}
+
+	balloonSizeMB := int64(float64(availableMemKB/1024) * .9)
+	if err := c.updateBalloon(ctx, balloonSizeMB); err != nil {
+		return status.WrapError(err, "inflate balloon")
+	}
+	if err := c.updateBalloon(ctx, 0); err != nil {
+		return status.WrapError(err, "deflate balloon")
+	}
+	return nil
+}
+
+// Best effort update the balloon to the target size.
+func (c *FirecrackerContainer) updateBalloon(ctx context.Context, targetSizeMib int64) error {
+	start := time.Now()
+	err := c.machine.UpdateBalloon(ctx, targetSizeMib)
+	if err != nil {
+		return err
+	}
+
+	var currentBalloonSize int64
+	defer func() {
+		log.CtxDebugf(ctx, "Update balloon to %d MB (target %d MB) took %s", currentBalloonSize, targetSizeMib, time.Since(start).String())
+	}()
+
+	// Wait for the balloon to reach its target size.
+	pollInterval := 300 * time.Millisecond
```

Comment:
would it make sense to do an exponential backoff here starting from like 5ms or something to reduce latency in case it gets updated very quickly? or does it not ever really take less than 300ms?

Reply:
The fastest I've seen it is ~1GB in 300ms and typically we'd be inflating the balloon over a GB